cmd/edenController.go

```diff
@@ -0,0 +1,463 @@
+package cmd
+
+import (
+	"fmt"
+	"github.com/docker/docker/pkg/fileutils"
+	"github.com/lf-edge/eden/pkg/utils"
+	"github.com/lf-edge/eve/api/go/config"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var controllerMode string
+var baseOSImage string
+var baseOSImageActivate bool
+var configItems map[string]string
+
+func getParams(line, regEx string) (paramsMap map[string]string) {
+
+	var compRegEx = regexp.MustCompile(regEx)
+	match := compRegEx.FindStringSubmatch(strings.TrimSpace(line))
+
+	paramsMap = make(map[string]string)
+	for i, name := range compRegEx.SubexpNames() {
+		if i > 0 && i <= len(match) {
+			paramsMap[name] = match[i]
+		}
+	}
+	return
+}
+
+func getControllerMode() (modeType, modeURL string, err error) {
+	params := getParams(controllerMode, controllerModePattern)
+	if len(params) == 0 {
+		return "", "", fmt.Errorf("cannot parse mode (not [file|proto|adam|zedcloud]://<URL>): %s", controllerMode)
+	}
+	ok := false
+	if modeType, ok = params["Type"]; !ok {
+		return "", "", fmt.Errorf("cannot parse modeType (not [file|proto|adam|zedcloud]://<URL>): %s", controllerMode)
+	}
+	if modeURL, ok = params["URL"]; !ok {
+		return "", "", fmt.Errorf("cannot parse modeURL (not [file|proto|adam|zedcloud]://<URL>): %s", controllerMode)
+	}
+	return
+}
+
+var controllerCmd = &cobra.Command{
+	Use:   "controller",
+	Short: "interact with controller",
+	Long:  `Interact with controller.`,
+}
+
+var edgeNode = &cobra.Command{
+	Use:   "edge-node",
+	Short: "manage EVE instance",
+	Long:  `Manage EVE instance.`,
+}
+
+var edgeNodeReboot = &cobra.Command{
+	Use:   "reboot",
+	Short: "reboot EVE instance",
+	Long:  `reboot EVE instance.`,
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		assingCobraToViper(cmd)
+		if baseOSVersionFlag := cmd.Flags().Lookup("os-version"); baseOSVersionFlag != nil {
+			if err := viper.BindPFlag("eve.base-tag", baseOSVersionFlag); err != nil {
+				log.Fatal(err)
+			}
+		}
+		_, err := utils.LoadConfigFile(configFile)
+		if err != nil {
+			return fmt.Errorf("error reading configFile: %s", err.Error())
+		}
+		return nil
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		modeType, modeURL, err := getControllerMode()
+		if err != nil {
+			log.Fatal(err)
+		}
+		log.Infof("Mode type: %s", modeType)
+		log.Infof("Mode url: %s", modeURL)
+		var changer configChanger
+		switch modeType {
+		case "file":
```

Comment:
So as per my comment above the right place to hook up the parsing of viper options is here and then we can pre-populate the *Changer context objects with what they have to know.